broadlink/climate.py

```diff
@@ -238,3 +239,333 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        and pad size.
+        """
+        length = 2 + len(payload)  # length prefix
+        packet_length = ((length - 1) // 16 + 1) * 16
+        packet = bytearray([length, 0])
+        packet.extend(payload)
+        packet.extend(self._calculate_checksum(packet[2:], 'little'))
+        packet.extend([0] * (packet_length - len(packet)))  # round size to 16
+
+        return packet
+
+    def _send_short_payload(self, payload: int) -> bytes:
+        """Send a request for info from AC unit and returns the response.
+        0 = GET_AC_INFO, 1 = GET_STATES, 2 = GET_SLEEP_INFO
+        """
+        packet = bytearray([0xbb, 0x00, 0x06, 0x80, 0x00, 0x00, 0x02, 0x00])
+        if (payload == 0):
+            packet.extend([0x21, 0x01])
+        elif (payload == 1):
+            packet.extend([0x11, 0x01])
+        elif (payload == 2):
+            packet.extend([0x41, 0x01])
+        # elif (payload == 3):
+        #     packet = bytearray(16)
+        #     packet[0x00] = 0xd0
+        #     packet[0x01] = 0x07
+        else:
+            raise ValueError(f"unrecognized payload type: {payload}")
+
+        response = self.send_packet(0x6a, self._encode(packet))
+        check_error(response[0x22:0x24])
+        return (self._decode(response))
+
+    def get_state(self) -> dict:
+        """Returns a dictionary with the unit's parameters.
+
+        Returns:
+            dict:
+                state (bool): power
+                target_temp (float): temperature set point 16<n<32
+                mode (str): cooling, heating, fan, dry, auto
+                speed (str): low, mid, high, auto
+                mute (bool):
+                turbo (bool):
+                swing_h (str): ON, OFF
+                swing_v (str): ON, OFF, 1, 2, 3, 4, 5 (fixed positions)
+                sleep (bool):
+                display (bool):
+                health (bool):
+                clean (bool):
+                mildew (bool):
+        """
+        payload = self._send_short_payload(1)
+        if (len(payload) != 32):
+            raise RuntimeError(f"unexpected payload size: {len(payload)}")
+
+        data = {}
+        data['state'] = payload[0x14] & 0x20 == 0x20
+        data['target_temp'] = (8 + (payload[0x0c] >> 3)
+                               + (0.0 if (payload[0xe] & 0b10000000) == 0 else 0.5))  # noqa E501
+
+        data['swing_h'] = {
+            0b000: 'ON',
+            0b111: 'OFF'
+        }.get((payload[0x0d] & 0b11100000) >> 5, 'unrecognized value')
+
+        data['swing_v'] = {
+            0b000: 'ON',
+            1: '1',
+            2: '2',
+            3: '3',
+            4: '4',
+            5: '5',
+            0b111: 'OFF'
+        }.get(payload[0x0c] & 0b111, 'unrecognized value')
+
+        data['mode'] = {
+            0x00: 'auto',
+            0x20: 'cooling',
+            0x40: 'drying',
+            0x80: 'heating',
+            0xc0: 'fan'
+        }.get(payload[0x11] &~ 0b111, 'unrecognized value')  # noqa E225
```

Comment:
My opinion on how to handle this like a boss. Import the IntEnum class and create child classes inside the controller class. So we would end up with something like this:
```python3
from enum import IntEnum, unique

class sq1:

    @unique
    class Mode(IntEnum):
        AUTO = 0
        COOLING = 0x20
        DRYING = 0x40
        HEATING = 0x80
        FAN = 0xc0
```

We can generate the mode in many different ways...
```python3
mode = self.Mode.FAN
mode = self.Mode["FAN"]
mode = self.Mode(0xc0)
```

We can use:
```python3
# For reading
mode = self.Mode(payload[0x11] &~ 0b111)

# For writing
packet[0xf] = mode << 3 | sleep << 2
```

This is like creating an interface for a normal integer.
```python3
# If we want to see the name
>>> mode.name
'COOLING'

# If we want to see the value
>>> mode.value
32
```
A ValueError is generated for invalid input:
```python3
>>> d = blk.hello("192.168.0.16")  # Example HVAC
>>> d.Mode(1000)
Traceback (most recent call last):
  File "<pyshell#59>", line 1, in <module>
    d.Mode(1000)
  File "/usr/lib/python3.6/enum.py", line 293, in __call__
    return cls.__new__(cls, value)
  File "/usr/lib/python3.6/enum.py", line 535, in __new__
    return cls._missing_(value)
  File "/usr/lib/python3.6/enum.py", line 548, in _missing_
    raise ValueError("%r is not a valid %s" % (value, cls.__name__))
ValueError: 1000 is not a valid Mode
```

External applications have access to an interface. They can see what options they have.
```python3
>>> d = blk.hello("192.168.0.16")  # Example HVAC
>>> [m for m in d.Mode.__members__]
['AUTO', 'COOLING', 'DRYING', 'HEATING', 'FAN']
```
Best of all: this is a normal integer that can be appended to the packet and serialized normally. What do you think?

Reply:
I like it! Will implement.
Would you then pass a Mode object in the dictionary as input or the string representation?